core/lib/Thelia/Controller/Admin/ProductController.php

```diff
@@ -1804,4 +1816,365 @@ protected function arrayHasEntries(array $data, array $entries)
 
         return $status;
     }
+
+    /*
+     * PRODUCT CLONING PROCESS
+     *
+     *  Vars prefixes :
+     * op = Original Product
+     * cp = Clone Product
+     */
+
+    /**
+     * @return mixed|\Symfony\Component\HttpFoundation\Response
+     * @throws \Exception
+     */
+    public function cloneAction()
+    {
+        if (null !== $response = $this->checkAuth($this->resourceCode, $this->getModuleCode(), [AccessManager::CREATE, AccessManager::UPDATE])) {
+            return $response;
+        }
+
+        // Initialize vars
+        $cpf = $this->createForm(AdminForm::PRODUCT_CLONE);
+        $lang = $this->getSession()->getLang()->getLocale();
+
+        try {
+            // Check the form against constraints violations
+            $form = $this->validateForm($cpf, "POST");
+
+            // Get the form field values
+            $data = $form->getData();
+
+            $opId = $data['productId'];
+
+            $cpId = $this->createCloneProduct($data, $lang, $opId);
+
+            $this->setCloneFeatureCombination($opId, $cpId);
+
+            $this->setProductAssociatedContent($opId, $cpId);
+
+            $this->setCloneI18n($lang, $opId, $cpId);
+
+            $this->setCloneImage($opId, $cpId);
+
+            $this->setCloneDocument($opId, $cpId);
+
```

Comment:
In all the above methods, you should use Thelia events instead of raw objects creation, as some modules (and future Thelia core components) may want to be notified when new objects are created, or changed.